File: TemplateMethodPattern/src/com/company/pavezimai/Pavezimas.java
package com.company.pavezimai;

import com.company.kainosSkaiciavimas.Ikainiai;
import com.company.kainosSkaiciavimas.IkainiaiBuilder;

import java.math.BigDecimal;

public abstract class Pavezimas {
    private final String vairuotojoVardas = "Vėjas Nupūstas";
    private final Ikainiai ikainiai = new IkainiaiBuilder()
            .setPradineKaina(100)
            .setKilometroKaina(100)
            .setMinutesKaina(12)
            .createIkainiai();

    private final int vairuotojasId;
    private final String isvykimoTaskas;
    private String atvykimoTaskas;
    private double atstumas;
    private double laikas;
    private BigDecimal kaina;

    public Pavezimas(int vairuotojasId, String isvykimoTaskas) {
        this.vairuotojasId = vairuotojasId;
        this.isvykimoTaskas = isvykimoTaskas;
        System.out.println("Jusu pavezimas pradetas, vairuotojas " + getVairuotojoVardas() + ".");
    }

    /**
     * @return Pavezimo kaina.
     */
    public BigDecimal uzbaigtiPavezima(double atstumas, double laikas, String atvykimoTaskas) throws UnsupportedOperationException {
        if (atvykimoTaskas.isEmpty()) {
            throw new UnsupportedOperationException("Nenustatytas atvykimo taskas.");
        }
        BigDecimal sumineKaina = apskaiciuotiKelionesKaina(atstumas, laikas, gautiVairuotojoIkainius());
        sumineKaina = koreguotiSumineKaina(getIsvykimoTaskas(), atvykimoTaskas, sumineKaina);
        System.out.println("Ačiū, kad važiavote su " + getVairuotojoVardas() + ".");
        this.kaina = sumineKaina;
        this.atstumas = atstumas;
        this.laikas = laikas;
        this.atvykimoTaskas = atvykimoTaskas;
        return sumineKaina;
    }

    public Ikainiai gautiVairuotojoIkainius() {
        return ikainiai;
    }

    public int getVairuotojasId() {
        return vairuotojasId;
    }

    public String getVairuotojoVardas() {
        return vairuotojoVardas;
    }

    public String getIsvykimoTaskas() {
        return isvykimoTaskas;
    }

    public String getAtvykimoTaskas() {
        return atvykimoTaskas;
    }

    public double getAtstumas() {
        return atstumas;
    }

    public double getLaikas() {
        return laikas;
    }

    public BigDecimal getKaina() {
        return kaina;
    }

    protected abstract BigDecimal koreguotiSumineKaina(String isvykimoTaskas, String atvykimoTaskas, BigDecimal sumineKaina);

    protected abstract BigDecimal apskaiciuotiKelionesKaina(double atstumas, double laikas, Ikainiai ikainiai);
}

File: TemplateMethodPattern/src/com/company/pavezimai/Pavezimas.scala
package com.company.pavezimai

import java.math.BigDecimal

import com.company.kainosSkaiciavimas.Ikainiai

abstract class Pavezimas(val vairuotojasId: Int, val isvykimoTaskas: String) {
  if (!(30000 < vairuotojasId && vairuotojasId < 40000)) throw new IllegalArgumentException("Nera vairuotojo su tokiu identifikaciniu numeriu.")
  val ikainiai: Ikainiai = nustatytiIkainius
  val vairuotojoVardas = "Vėjas Nupūstas"
  var atvykimoTaskas: String = _
  var atstumas = .0
  var laikas = .0
  var kaina: BigDecimal = _
  System.out.println("Jusu pavezimas pradetas, vairuotojas " + vairuotojoVardas + ".")

  private def nustatytiIkainius = new Ikainiai(100, 100, 12, 0)

  /**
   * @return Pavezimo kaina.
   */
  @throws[UnsupportedOperationException]
  def uzbaigtiPavezima(atstumas: Double, laikas: Double, atvykimoTaskas: String): BigDecimal = {
    if (atvykimoTaskas.isEmpty) throw new UnsupportedOperationException("Nenustatytas atvykimo taskas.")
    var sumineKaina: BigDecimal = this.apskaiciuotiKelionesKaina(atstumas, laikas, ikainiai)
    sumineKaina = this.koreguotiSumineKaina(isvykimoTaskas, atvykimoTaskas, sumineKaina)
    System.out.println("Ačiū, kad važiavote su " + vairuotojoVardas + ".")
    this.kaina = sumineKaina
    this.atstumas = atstumas
    this.laikas = laikas
    this.atvykimoTaskas = atvykimoTaskas
    sumineKaina
  }

  protected def koreguotiSumineKaina(isvykimoTaskas: String, atvykimoTaskas: String, sumineKaina: BigDecimal): BigDecimal

  protected def apskaiciuotiKelionesKaina(atstumas: Double, laikas: Double, ikainiai: Ikainiai): BigDecimal
}